Feat: Caret API (#41084)
mdn/content: files/en-us/web/css/caret-shape/index.md

sidebar: cssref
---

{{SeeCompatTable}}

The **`caret-shape`** [CSS](/en-US/docs/Web/CSS) property sets the shape of the **insertion caret**, the visible marker that appears in editable elements to indicate where the next character will be inserted or deleted.

{{InteractiveExample("CSS Demo: caret-shape")}}

```css interactive-example-choice
caret-shape: auto;
```

```css interactive-example-choice
caret-shape: bar;
```

```css interactive-example-choice
caret-shape: block;
```

```css interactive-example-choice
caret-shape: underscore;
```

```html interactive-example
<section class="default-example container" id="default-example">
  <div>
    <p>Click in the field to see the caret shape:</p>
    <p><input id="example-element" type="text" value="Sample text" /></p>
  </div>
</section>
```

```css interactive-example
#example-element {
  font-size: 1.2rem;
  padding: 8px;
  width: 300px;
}
```

## Syntax

```css
/* Keyword values */
caret-shape: auto;
caret-shape: bar;
caret-shape: block;
caret-shape: underscore;

/* Global values */
caret-shape: inherit;
caret-shape: initial;
caret-shape: revert;
caret-shape: revert-layer;
caret-shape: unset;
```

### Values

- `auto`

>   - : The browser determines the caret shape. This typically follows platform conventions and may change based on context (e.g., insert mode vs. overtype mode).

Reviewer: ```suggestion
  - : The browser determines the caret shape. This typically follows platform conventions and may change based on context, for example, insert mode vs. overtype mode.
```

I'm just thinking, does everyone know what insert mode and overtype mode are? You can infer the meaning, but it might be worth a sentence to make it clear what they are, and say what caret shape is typically used for what mode. UPDATE: put this in the Description?

- `bar`
  - : The caret appears as a thin vertical line at the insertion point, positioned between characters rather than over them.

- `block`
  - : The caret appears as a rectangle that overlaps the next character after the insertion point. If no character follows, it appears after the last character.

- `underscore`
  - : The caret appears as a thin horizontal line underneath the next character after the insertion point. If no character follows, it appears after the last character.

## Description

> The insertion caret is the blinking cursor that indicates where text will be inserted when typing. Different caret shapes can provide visual feedback about the current editing mode or simply offer aesthetic customization.

Reviewer: ```suggestion
The insertion caret is the blinking cursor that indicates where text will be inserted when typing. Different caret shapes can provide visual feedback about the current editing mode or offer visual customization.
```

On MDN, we are against the use of "simple"/"simply" ;-)

Also, I wonder if "aesthetic" is a suitable word for an international audience; would "visual" serve the same purpose but be a bit more obvious?

Author: Noted!

### Caret positioning and behavior

The `caret-shape` property affects how the caret is visually rendered but doesn't change its logical position in the text. The caret always represents the insertion point between characters.

> - **Bar carets** are positioned between characters and are most common in modern interfaces

Reviewer: Put a full stop at the end of each bullet.

> - **Block carets** overlay the next character and are often used in terminal applications or to indicate overtype mode

Reviewer: Hrm, actually, the explanation of insert versus overtype mode is probably better put in the description.

> - **Underscore carets** appear below characters and can be useful for certain design aesthetics

Reviewer: Such as?

### Interaction with writing modes

> The caret shape adapts to the writing mode of the text. In vertical writing modes, bar carets become horizontal, and underscore carets position themselves appropriately relative to the text direction.

Reviewer: ```suggestion
The caret shape adapts to the {{cssxref("writing-mode")}} of the text. In vertical writing modes, bar carets become horizontal, and underscore carets position themselves appropriately relative to the text direction.
```

## Formal definition

{{cssinfo}}

## Formal syntax

{{csssyntax}}

> ## Examples

Reviewer: OK, this is really weird. The live preview is showing three examples (which are all really good, BTW!) but the PR is only showing one!

In general, they all look fine, although generally on MDN we like the explanation to be before the code blocks rather than in comments inside the code blocks. 

Also, you might want to make sure that:

- The embed iframes are set tall enough so that you don't get scroll bars on the examples.
- The examples are centered so that they don't stick to the sides of the embed frame (the terminal one would look better centered)
- You provide more descriptive text for each option. Having "Click here" written four times is not good for accessibility. Something that varies like "Edit me to see the bar caret" would be better.

### Terminal-style interface

This example shows how `caret-shape: block` can be used to create a terminal or command-line interface appearance.

#### HTML

```html
<div class="terminal">
  <div class="terminal-header">Terminal</div>
  <div class="terminal-content">
    <div class="prompt">
      user@computer:~$ <input type="text" class="terminal-input" />
    </div>
  </div>
</div>
```

#### CSS

```css hidden
.terminal {
  background-color: #1e1e1e;
  color: #00ff00;
  font-family: "Courier New", monospace;
  border-radius: 8px;
  overflow: hidden;
  width: 500px;

>   margin: 20px 0;

Reviewer: ```suggestion
  margin: 20px auto;
```

Center it?

}

.terminal-header {
  background-color: #333;
  padding: 8px 12px;
  font-size: 0.9rem;
  border-bottom: 1px solid #555;
}

.terminal-content {

>   padding: 12px;

Reviewer: Maybe set a height on it, like 100px or something? Terminals are not usually shown with one line of content.

}

.prompt {
  display: flex;
  align-items: center;
}
```

```css
.terminal-input {
  background: transparent;
  border: none;
  color: #00ff00;
  font-family: inherit;
  font-size: 1rem;
  outline: none;
  flex: 1;
  /* Block caret for terminal appearance */
  caret-shape: block;
  caret-color: #00ff00;
}
```

#### Result

{{EmbedLiveSample('Terminal-style_interface', 550, 120)}}

## Specifications

> {{Specifications}}

Reviewer: I'm assuming the spec and browser compat data is being added to BCD?

Author: yes!

#### HTML

```html
<div class="old-screen">
  <span>></span>
  <textarea class="terminal-input" placeholder="Enter command..."></textarea>
</div>
```

#### CSS

```css hidden
.old-screen {
  background: repeating-linear-gradient(
    #092104,
    #092104 2px,
    #123208 2px,
    #123208 4px
  );
  height: 140px;
  border-radius: 25px;
  display: flex;
  align-items: flex-start;
  padding: 20px;
  font-family: monospace;
}

span {
  display: inline-block;
  padding: 2px 5px;
  color: green;
  font-weight: bold;
  margin-right: 8px;
}

.terminal-input {
  background: transparent;
  height: 100%;
  border: none;
  color: green;
  font-family: inherit;
  font-size: 1rem;
  outline: none;
  flex: 1;
  resize: none;
}

.terminal-input::placeholder {
  color: #4a6741;
}
```

```css
.terminal-input {
  caret-shape: block;
  caret-animation: manual;
  animation: old-caret 2s infinite;
}

@keyframes old-caret {
  from,
  50% {
    caret-color: green;
  }
  75%,
  to {
    caret-color: transparent;
  }
}
```

#### Result

{{EmbedLiveSample('Retro_terminal_with_animated_caret', 550, 180)}}

### Console interface with underscore caret

> This example demonstrates using `caret-shape: underscore` to create a console-style interface where the underscore caret complements the terminal aesthetic.

Reviewer: These examples are so cool! I would definitely suggest getting rid of the rounded corners and placeholder for this one. You don't get a placeholder like that in a terminal

#### HTML

```html
<div class="console-demo">
  <div class="console-output">
    <p>user@host:css-ui-4 $ ls -a</p>
    <p>. .. Overview.bs Overview.html</p>
  </div>
  <div class="console-input">
    <span class="prompt">user@host:css-ui-4 $ </span>
    <input type="text" class="console" placeholder="Enter command..." />
  </div>
</div>
```

#### CSS

```css hidden
.console-demo {
  background: black;
  color: white;
  font-family: monospace;
  padding: 20px;
  height: 60px;
  border-radius: 4px;
}

.console-output {
  margin-bottom: 0.5rem;
}

.console-output p {
  margin: 0 0.25rem;
}

.console-input {
  display: flex;
  align-items: center;
}

.prompt {
  margin-right: 0;
}

.console {
  background: transparent;
  border: none;
  color: white;
  font-family: inherit;
  outline: none;
  flex: 1;
}
```

```css
.console {
  caret-shape: underscore;
}
```

#### Result

{{EmbedLiveSample('Console_interface_with_underscore_caret', 550, 100)}}

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - {{cssxref("caret-color")}}, {{cssxref("caret-animation")}}, {{cssxref("caret")}}

Reviewer: ```suggestion
- {{cssxref("caret-color")}}, {{cssxref("caret-animation")}}
- {{cssxref("caret")}} shorthand
```

Ooops, sorry, we usually put shorthand on a separate line to the longhands.

Reviewer: Also, maybe while you are working on this, you could have a look at the caret-color and caret-animation ref pages and make sure they follow the same pattern in their See also?

Reviewer: I've just thought of something else as well — you need to update https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_basic_user_interface, change the line that begins "The CSS basic user interface module also defines the `caret`, `caret-shape`, ..." to remove `caret-shape`, and add `caret-shape` to the properties list.

You could also consider turning `caret` into a link to that page, but I'd leave it where it is for now because no browser supports it yet.
